### ISSUE TITLE
Freezes When Trying to Display Error Dialog For Missing Images

### ISSUE THREAD

**Issue by General-Dru-Zod** [triplea-game/triplea#2820]
### My Operating System: 
Windows 10

### Engine version:
8281, 8284

### Map name and version:
Big World 3 Test Mod, 
Not uploaded yet. But if you need it I can attach it.

### Can you describe how to trigger the error? (eg: what sequence of actions will recreate it?)
I try to play game locally and it consistently freezes during game load. I have no errors displayed and console is also frozen, if I use the always display console feature. But also freezes with that feature off. I have a screenshot of half loaded map, and same image every time I try. 

I have no problems loading the game with 8131 even though it displays some errors related to missing images, which was expected since I was testing changes to xml.

It seems to freeze up at the moment the missing image errors normally pop up in 8131. I am attaching that error. But as I mentioned that error was expected.

### Do you have the exact error text? Please copy/paste if so
No errors, simply freezes. The console is also frozen, it displays normal loading locations but didn't get to missing images before it froze.
### Instead of this error, what should have happened?
Should load and display the expected missing image errors.
### Any additional information that may help:
[8281.Freezes.On.Load_01.07.2018.zip](https://github.com/triplea-game/triplea/files/1609874/8281.Freezes.On.Load_01.07.2018.zip)

**Comment by ron-murhammer** [comment]
@General-Dru-Zod Can you provide a link to the map zip and a save game so we can reproduce it.

**Comment by General-Dru-Zod** [comment]
Here is the link to zip. Its the Test Mod version of Big World 3 : FS that freezes. There is now save game just try load fresh game of it. 

Disclaimer this Test Mod version isn't meant for true play. However should still be able to load to test xml changes before implementing them in actual.

https://www.sendspace.com/file/x1hs4z

**Comment by ron-murhammer** [comment]
@General-Dru-Zod Thanks. I was able to reproduce the freeze as well and it appears to be due to some recent changes around showing error dialog instead of console. 

@DanVanAtta Most likely the issue is due to the recent changes in #2782. The game freezes when trying to buildAndShow the error dialog at https://github.com/triplea-game/triplea/blob/master/src/main/java/games/strategy/debug/ClientLogger.java#L103

**Comment by ssoloff** [comment]
I may be doing something wrong, but to download the Big World 3 map, I'm being prompted to sign up for a "free" account where I have to provide a credit card number.  Ain't no way I'm doing that on a site where the English spelling and grammar resemble a Nigerian prince email. :smile:

A simple repro step is to delete a common image from the _assets/units/_ folder.  I deleted _assets/units/Russians/infantry.png_ and started a `world_war_ii_classic` map to repro the freeze.

**Comment by ssoloff** [comment]
A bit more info...

There's a deadlock between the EDT and the "Map panel background drawer" (I'll call it MPBDT below) threads.  The MPBDT acquires a `Tile` lock, then encounters the missing image error, and calls `ClientLogger#logError()` while the `Tile` lock is still held.  At this time, the EDT is trying to acquire the same `Tile` lock.  Because `ErrorMessageBuilder` does an `invokeAndWait()`, thte MPBDT cannot make any progress because it's waiting on the EDT lock.

Obviously, the MPBDT shouldn't be calling out to other threads while it's holding a lock.  The following naive solution prevents the freeze, but results in multiple error dialogs being displayed simultaneously (in my case, one per Russian infantry unit at the start of the game):

```diff
diff --git a/src/main/java/swinglib/ErrorMessageBuilder.java b/src/main/java/swinglib/ErrorMessageBuilder.java
index 8fa268d32..7fffba5f5 100644
--- a/src/main/java/swinglib/ErrorMessageBuilder.java
+++ b/src/main/java/swinglib/ErrorMessageBuilder.java
@@ -8,12 +8,11 @@ import java.util.ArrayList;
 import java.util.List;
 
 import javax.swing.JOptionPane;
+import javax.swing.SwingUtilities;
 
 import com.google.common.base.Preconditions;
 import com.google.common.base.Strings;
 
-import games.strategy.ui.SwingAction;
-
 
 /**
  * Swing builder for an error message pop-up (Uses JOptionPane). Example usages:
@@ -64,7 +63,7 @@ public class ErrorMessageBuilder {
    * Values that must be set: title, contents
    */
   public void buildAndShow() {
-    SwingAction.invokeAndWait(this::buildAndShowSwingSafe);
+    SwingUtilities.invokeLater(this::buildAndShowSwingSafe);
   }
 
   private void buildAndShowSwingSafe() {
```

I guess the nice thing about the legacy Error Console behavior was that it could handle concurrent errors asynchronously.  The new dialog-based approach is inherently synchronous and may require a bit of a redesign to handle this type of scenario.

**Comment by ron-murhammer** [comment]
@ssoloff Yeah, I kind of figured it was a thread deadlock. Given that concurrent errors are pretty common (usually your missing multiple images not just one), it might be better to just roll back this change until it can be further flushed out and tested.

**Comment by ssoloff** [comment]
@ron-murhammer I submitted #2822 to temporarily fix the problem on `master`.  I'm going to assign this issue to @DanVanAtta and leave it open for him to consider a redesign that will allow us to retain the functionality while fixing the potential for deadlock described above.  If no progress is made in a week or so, I'll submit a follow-up PR to revert the feature.

**Comment by ron-murhammer** [comment]
@ssoloff Thanks. I'm good with that solution for now as it restores functionality and at least gets rid of the deadlock.

**Comment by General-Dru-Zod** [comment]
@ssoloff 
Ah sorry just saw this. Regarding the dl of big world 3. You can just x out the signup page and there should be a page behind that, which allows you to dl the map. But looks like you guys resolved it anyways.

**Comment by DanVanAtta** [comment]
Labels.. could use work.

`P1`: impacting current players, problem is in the latest release
`regression`: loss of functionality compared to latest release

Before mitigation, this was a `regression`, before the game would launch, after introducing a bug it no longer launched.

After mitigation, we have no bugs here anymore, the code is disabled - `type: bug` is not appropriate. We do on the other hand have technical debt to clean up the disabled code and fix it, so `category: code` fits.
-------------------------

With regards to the original problem, not showing it as a pop-up will fix the issue: https://github.com/triplea-game/triplea/pull/2830

Bit of shame though that the missing unit image is written quietly to the console logger and would not be a pop-up, map makers will need to remember to open it to verify it is clean (which they kinda need to do anyways).

**Comment by DanVanAtta** [comment]
Considering bug is gone, this issue can be closed. It was an extremely useful bug report, I don't know how long before maybe this issue would have been caught.

To track the disabled "better error handling", I created ticket: https://github.com/triplea-game/triplea/issues/2831

**Comment by ssoloff** [comment]
> Bit of shame though that the missing unit image is written quietly to the console logger and would not be a pop-up, map makers will need to remember to open it to verify it is clean (which they kinda need to do anyways).

They can also use the new "Show Console Always" setting on the Test tab, which, ironically, was requested by a map maker. :smile:
